llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -15527,6 +15532,28 @@ SDValue DAGCombiner::visitTRUNCATE(SDNode *N) {
     break;
   }
 
+  if (TLI.isNarrowingProfitable(N0.getNode(), SrcVT, VT)) {
+    switch (N0.getOpcode()) {
+    case ISD::ADD:
+    case ISD::SUB:
+    case ISD::MUL:
+    case ISD::AND:
+    case ISD::OR:
+    case ISD::XOR:
+      if (!(N0.hasOneUse() && VT.isScalarInteger()))
+        break;
+      if (LegalOperations && !TLI.isOperationLegal(N0.getOpcode(), VT))
+        break;
```

Comment:
Swap this check with the isNarowingProfitable above? This is the simpler condition